packages/graphql-armor/src/apollo/armor.ts

```diff
@@ -24,26 +27,38 @@ export class ApolloArmor {
   }
 
   protect(): {
-    plugins: PluginDefinition[];
+    plugins: any;
```

Comment:
I think you can use `ApolloServerPlugin` from `@apollo/server` now

Reply:
@michael-watson Yep, that was actually my [initial implementation](https://github.com/Escape-Technologies/graphql-armor/pull/354/commits/639546eeec2fc2c3c096e3b621c9a32b6d494e41#diff-ce8dd078c3970c54e30c55f80357edb7d6445381483997fbaaf7f98856362182R1), but that is not backwards compatible with `Apollo Server 3`, and having TS issues so I had to put `any`. If it's not `any`, it will result to this in a project still using `Apollo Server 3`.

<img width="1449" alt="Screenshot 2023-03-14 at 6 17 42 AM" src="https://user-images.githubusercontent.com/653750/224849259-f581c8f8-c645-4588-b029-f99fa6502a8d.png">

Reply:
```suggestion
    plugins: ApolloServerOptions<BaseContext>['plugins'];
```

would that works?

Reply:
Hello @c3b5aw,

That change would cause issues to both AS3 and AS4 default examples like so:

Apollo Server 4
<img width="1010" alt="Screenshot 2023-03-14 at 7 33 30 AM" src="https://user-images.githubusercontent.com/111724678/224855013-63c9eaf0-0229-4cee-b2b5-04a2d2a1bfdf.png">

Apollo Server 3
<img width="1062" alt="Screenshot 2023-03-14 at 7 33 24 AM" src="https://user-images.githubusercontent.com/111724678/224855020-0cbee215-8032-4ffe-b632-0d7393f7da66.png">

Reply:
Hello @c3b5aw ,

That change to `plugins: ApolloServerOptions<BaseContext>['plugins'];` would cause issues to both AS4 and AS3 default examples like so:

Apollo Server 4
<img width="1010" alt="Screenshot 2023-03-14 at 7 33 30 AM" src="https://user-images.githubusercontent.com/653750/224855566-bc819d43-8a82-421b-8efe-820865117ebb.png">

Apollo Server 3
<img width="1062" alt="Screenshot 2023-03-14 at 7 33 24 AM" src="https://user-images.githubusercontent.com/653750/224855568-7b9ddbbb-28d6-4ae9-8752-03f9431a8819.png">

Reply:
> 

You could get back to it.

`@apollo/server` has been pushed through renovate.

cf: #358 #359 

Do we have any other alternative ? I want to avoid any as much as possible.

Reply:
Though we can duplicate that code until apollo v3 is deprecated

Reply:
@michael-watson @c3b5aw 

Upon checking the difference between the `ApolloServerPlugin`, I am not seeing how this is possible since they differ in `serverWillStart` type definition. 

I think AS4 is a major rewrite when I compare the difference between interfaces and the attribute types and unless AS3 packages are reinstalled back to armor and the plugins will have a `plugins: PluginDefinition[] | PluginAS3Definition[]` where it handles both, AS3 will not be compatible if not using `plugins: any`.

The purpose of the PR was to eliminate AS3 related packages for those that already migrated to AS4 😅 so I think either use `any` for `plugins` or revert to a separate class for AS4 e.g `ApolloV4Armor` use PluginDefinition there and use any for `ApolloArmor`.

Otherwise, please feel free to modify this PR should you find a way to make it work with AS3 without reintroducing AS3 packages.

ApolloServerPlugin - AS3 uses `service: GraphQLServiceContext,`
```js
export interface ApolloServerPlugin<
  TContext extends BaseContext = BaseContext,
> {
  serverWillStart?(
    service: GraphQLServiceContext,
  ): Promise<GraphQLServerListener | void>;

  requestDidStart?(
    requestContext: GraphQLRequestContext<TContext>,
  ): Promise<GraphQLRequestListener<TContext> | void>;
}
```

ApolloServerPlugin - AS4 uses `service: GraphQLServerContext,`
```js
export interface ApolloServerPlugin<
  in TContext extends BaseContext = BaseContext,
> {
  // Called once on server startup, after the schema has been loaded.
  serverWillStart?(
    service: GraphQLServerContext,
  ): Promise<GraphQLServerListener | void>;

  // Called once per request, before parsing or validation of the request has
  // occurred. This hook can return an object of more fine-grained hooks (see
  // `GraphQLRequestListener`) which pertain to the lifecycle of the request.
  requestDidStart?(
    requestContext: GraphQLRequestContext<TContext>,
  ): Promise<GraphQLRequestListener<TContext> | void>;

  /**
   * "Unexpected" errors do not include more common errors like validation,
   * parsing, or graphql execution errors. Rather, an unexpected error might
   * occur when a plugin hook throws unexpectedly or a bug in Apollo Server is
   * encountered. Notably, when errors like this occur, the error is masked to
   * the client.
   */
  unexpectedErrorProcessingRequest?({
    requestContext,
    error,
  }: {
    requestContext: GraphQLRequestContext<TContext>;
    error: Error;
  }): Promise<void>;
  // Called specifically when the user-provided `context` function throws an
  // error.
  contextCreationDidFail?({ error }: { error: Error }): Promise<void>;
  /**
   * This hook is called any time a "Bad Request" error is thrown during request
   * execution. This includes CSRF prevention and malformed requests (e.g.
   * incorrect headers, invalid JSON body, or invalid search params for GET),
   * but does not include malformed GraphQL.
   */
  invalidRequestWasReceived?({ error }: { error: Error }): Promise<void>;
  // Called on startup fail. This can occur if the schema fails to load or if a
  // `serverWillStart` or `renderLandingPage` hook throws.
  startupDidFail?({ error }: { error: Error }): Promise<void>;
}
```

---
GraphQLServerContext vs GraphQLServiceContext
```js
export interface GraphQLServerContext {
  readonly logger: Logger;
  readonly cache: KeyValueCache<string>;

  schema: GraphQLSchema;
  apollo: ApolloConfig;
  startedInBackground: boolean;
}
```
```js
export interface GraphQLServiceContext {
  logger: Logger;
  schema: GraphQLSchema;
  /**
   * @deprecated: a not particularly stable or useful hash of the schema.
   */
  schemaHash: SchemaHash;
  apollo: ApolloConfig;
  persistedQueries?: {
    cache: KeyValueCache;
  };
  serverlessFramework: boolean;
}
```

Reply:
Alright @arvi,

Thanks a lot for your input.
Today, I merged the new apollo server packages, I'll go back to `ApolloServerPlugin`.